[RFC] man.vim: default mapping and general improvements

nhooyr opened the pull request:
See https://github.com/neovim/neovim/pull/4449#issuecomment-237809542

All I've done thus far is make `<Plug>(Man)` use `<cword>` for non-man buffers (and some small improvements to other parts of the code), just like `keywordprg` because the syntax is unpredictable. But in man buffers, we use `<cWORD>`. I'm not sure how to continue because I am very unfamiliar with c and it looks like the default `K` mapping is hard coded in c.

Also, was man.vim the only reason `keywordprg` accepted commands? If so, I think we should remove that behaviour.

nhooyr committed:
man.vim: default K mapping

- Also some small improvements in other parts.

nhooyr commented:
Actually, I'm not so sure about using `<cword>` for other filetypes.
Say there is a comment in a C file with the manpage refrence:`printf(3)`. If you press `K` on it, it will actually take you to `printf` because `expand('<cword>')` returned `printf`. And if you're on the 3, it will take you nowhere and print a not found error.

justinmk commented:
> was man.vim the only reason keywordprg accepted commands? If so, I think we should remove that behaviour.

Why?

nhooyr commented:
Alright, forget all that, I'm very sleepy :(. I've made the command work with `keywordprg`. Turns out, `v:count1` is available for commands as well so I can use the same trick I used with the mappings to ensure the user explicitly set the section.

When you open a man buffer though, the mapping for `K` switches to the one that uses `<cWORD>`. This works better for manpages because I'd rather not try and compile a list of allowable characters in manpage names.

This way, the current defaults work perfectly fine with minimal change.

justinmk commented on runtime/plugin/man.vim:
```diff
@@ -5,6 +5,12 @@ if exists('g:loaded_man')
+      \ if exists('b:manwidth') && line('$') == 1 && empty(getline(1)) |
+      \   unlet b:manwidth |
+      \   silent execute 'Man' matchstr(expand('<amatch>'), 'man://\zs.*') |
+      \ endif
```
This seems like a hack. Why is the buffer empty?

nhooyr replied:
I'm gonna actually leave this out for another PR.

tweekmonster replied:
@justinmk It was a suggestion I made in #5169.  Out of habit, I use `:e` to reload a buffer.  For me, it made sense to do that when I resized the window and wanted the man page to be re-rendered.

nhooyr committed:
man.vim: refactoring and autocmd fix

- man#open_page_command and man#open_page_mapping are now a single
  function
- New autocmd to fix #5172

nhooyr committed:
man.vim: buffers are now listed

- Since the names are set and ':vsplit printf(3)' work, there is no need
to unlist them.

nhooyr commented:
For posterity, https://github.com/nhooyr/neovim/commit/f34319df032ba35075b4410287c99df0ee29d651#commitcomment-18549463

justinmk commented:
re https://github.com/nhooyr/neovim/commit/f34319df032ba35075b4410287c99df0ee29d651#commitcomment-18549824

- Help files are not restored in sessions, unless a flag is set. 
- It makes sense to list manpage buffers for  `:ls`, `:bnext`, etc.
- With `buftype=` , what happens when you do `:confirm qa` ? Does it confirm closing the manpage? 

I think not supporting sessions is the right thing to do for now. Let's try to fix the general problem of why sessions aren't working with `buftype=` instead of making man.vim have incorrect settings. It may requires changing the session restore logic, even. (Separate PR, of course)

nhooyr commented:
No it does not confirm closing the manpage.

Alright, now `buftype=nofile` is set and the buffers are listed.
